### PR TITLE
Remove Forms and simplify the Input API.

### LINKED ISSUE #6018: Input should either assert or document that InputValue needs to be persistent.
It's currently not obvious that `InputValue`s need to be preserved in a state.

### PR DESCRIPTION
**Pull request by mpcomplete**
Input now accepts its text, selection, and composing arguments
separately. Any can be null, meaning the Input widget will manage them
internally.

Form provides very little added value beyond what Input widget now does,
so I've removed it to simplify the API.

Fixes https://github.com/flutter/flutter/issues/6018 and https://github.com/flutter/flutter/issues/6097

### CODE REVIEW

File: examples/flutter_gallery/lib/demo/text_field_demo.dart [hunk 1]
```diff
@@ -65,40 +65,35 @@ class TextFieldDemoState extends State<TextFieldDemo> {
...
   Widget build(BuildContext context) {
     return new Scaffold(
       key: _scaffoldKey,
       appBar: new AppBar(
         title: new Text('Text fields')
       ),
-      body: new Form(
-        child: new Block(
+      body: new Block(
           padding: const EdgeInsets.symmetric(horizontal: 16.0),
```
Comment (abarth) [inline examples/flutter_gallery/lib/demo/text_field_demo.dart]: Looks like the indent is off now.

File: packages/flutter/lib/src/material/input.dart [hunk 2]
```diff
@@ -65,7 +66,11 @@ class Input extends StatefulWidget {
   ///
   /// Otherwise, this is the current value, and must be updated every
   /// time [onChanged] is called.
-  final InputValue value;
+  // final InputValue value;
+
+  final String text;
+  final TextSelection selection;
+  final TextRange composing;
```
Comment (abarth) [inline packages/flutter/lib/src/material/input.dart]: These need docs.

File: packages/flutter/lib/src/material/input.dart [hunk 3]
```diff
@@ -65,7 +66,11 @@ class Input extends StatefulWidget {
   ///
   /// Otherwise, this is the current value, and must be updated every
   /// time [onChanged] is called.
-  final InputValue value;
+  // final InputValue value;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/input.dart]: commented out code

**Comment by Hixie** [comment]
One issue with this is that there's no good way to reset a text field. Suppose the form got committed or something, and you want to change the contents of the input, blow away the selection, blow away the composing range, blow away the voice recognition context state, etc. You have to know everything that InputValue has to do this. There's no way to just blow it all away without losing the focus state (i.e. without changing the key on the Input).

One option would be to take an InputValue instead of the explicit text/selection/composition range fields, and allow that input value to have null fields, and have a "merge" method on the input value that you apply instead of having the logic in _setCurrentValue that takes all the non-null fields of the new value and obliterates the corresponding values on the old object, leaving any fields of the old object that are null in the new object unchanged.

**Comment by mpcomplete** [comment]
> One issue with this is that there's no good way to reset a text field. Suppose the form got committed or something, and you want to change the contents of the input, blow away the selection, blow away the composing range, blow away the voice recognition context state, etc. You have to know everything that InputValue has to do this. There's no way to just blow it all away without losing the focus state (i.e. without changing the key on the Input).

You can set any of the 3 fields on Input, so to blow them away, just pass in `text: '', selection: emptySelection, composing: TextRange.empty`. Or did you mean something else?

> One option would be to take an InputValue instead of the explicit text/selection/composition range fields, and allow that input value to have null fields, and have a "merge" method on the input value that you apply instead of having the logic in _setCurrentValue that takes all the non-null fields of the new value and obliterates the corresponding values on the old object, leaving any fields of the old object that are null in the new object unchanged.

This is along the lines of the approach I took here: https://github.com/flutter/flutter/pull/6460 . That also seems OK to me, though it might be a little more confusing that omitting a field from InputValue means the Input widget manages it for you.

**Comment by Hixie** [comment]
> You can set any of the 3 fields on Input, so to blow them away, just pass in...

How do you know those 3 fields are enough? One day we'll add more (e.g. maybe the voice recognition context I mentioned).

**Comment by mpcomplete** [comment]
> How do you know those 3 fields are enough? One day we'll add more (e.g. maybe the voice recognition context I mentioned).

That's a good point, I missed that. And after updating some other examples, it's definitely more cumbersome to have to provide 3 fields.

I'm thinking the single InputValue field with nullable components is the best approach. @abarth didn't like that approach in https://github.com/flutter/flutter/pull/6460 , though.

One solution halfway between the two: if you provide a new `text` value, the underlying `Input`/`RawInput` resets the other InputValue fields for you (which is kind of required anyway, since the selection may not make sense with the new text). Then in the common case, you're only managing the `text` field.

**Comment by mpcomplete** [comment]
OK, I've implemented my "halfway" solution. If you provide a new text value, `RawInput` will clear the other fields. https://github.com/flutter/flutter/pull/6533/files#diff-4b2d1cad6cad02a17da658ceab24b3f8R252

It feels pretty good in the common case (the user doesn't need to manually control the selection or composing ranges). I'm also happy to change it back to accepting an InputValue whose fields can be null.

Either way seems like an improvement over the old Form+Input confusion.

File: packages/flutter/lib/src/material/input.dart [hunk 4]
```diff
@@ -54,18 +51,21 @@ class Input extends StatefulWidget {
...
 
-  /// The text of the input field.
-  ///
-  /// If the [Input] is in a [Form], this is the initial value only.
-  ///
-  /// Otherwise, this is the current value, and must be updated every
-  /// time [onChanged] is called.
-  final InputValue value;
+  /// The editable text displayed in the input field. If this is null, the
+  /// [Input] manages it for you.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/input.dart]: this should discuss all the implications of setting it and not (e.g. that setting it to a value that doesn't match the state causes the selection to be blown away, etc)

File: packages/flutter/lib/src/material/input.dart [hunk 5]
```diff
@@ -54,18 +51,21 @@ class Input extends StatefulWidget {
...
+  final String text;
+
+  /// The range of editable text that is currently selected. If this is null,
+  /// the [Input] manages it for you.
+  final TextSelection selection;
+
+  /// The range of editable text that is still being composed. If this is null,
+  /// the [Input] manages it for you.
+  final TextRange composing;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/input.dart]: Shouldn't these two be replaced by an InputValue? Maybe I misunderstood your "half-way" proposal. So long as you have to call out each part of the InputValue separately, the problem still exists, no?

File: packages/flutter/lib/src/widgets/editable.dart [hunk 6]
```diff
@@ -219,19 +228,30 @@ class RawInputState extends ScrollableState<RawInput> implements TextInputClient
   @override
   BoundedBehavior get scrollBehavior => super.scrollBehavior;
 
+  void _setCurrentValue({String text, TextSelection selection, TextRange composing}) {
+    InputValue updatedValue = new InputValue(
+      text: text ?? _currentValue?.text ?? '',
+      selection: selection ?? _currentValue?.selection ?? const TextSelection.collapsed(offset: -1),
+      composing: composing ?? _currentValue?.composing ?? TextRange.empty
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/editable.dart]: fundamentally i think this is the statement that's the problem

**Comment by Hixie** [comment]
I think your last change just reverses the issue. Now, if I want to maintain the selection when I change the text, I have to know all the fields I need to keep track of.

**Comment by mpcomplete** [comment]
> Now, if I want to maintain the selection when I change the text, I have to know all the fields I need to keep track of.

Do you? If you change the text and the selection, the `composing` is reset and then managed by the Input. That seems ok.

Do you prefer passing a single InputValue with fields that can be null, a la https://github.com/flutter/flutter/pull/6460 ?

**Comment by Hixie** [comment]
> Do you? If you change the text and the selection, the `composing` is reset and then managed by the Input. That seems ok.

Why would you want the `composing` to be reset though? (And the voice recognition context, and whatever else?) For example, suppose you just have a feature that increments the first number in a text field by one. Why would doing that also reset the `composing`?

> Do you prefer passing a single InputValue with fields that can be null, a la https://github.com/flutter/flutter/pull/6460 ?

I think that's less likely to break in the future.

I think Adam's comment in that PR isn't specifically against doing that, by the way. I think he's just saying that he wishes we had a solution that worked better at the global level, e.g. more similar to what you were talking about with grabbing the setter and so on. This PR has the same issue as that one. (I think that's fine, incremental improvements are still improvements, and may lead us to finding inspiration for a way that addresses the wider issue.)

**Comment by Hixie** [comment]
@mpcomplete Does #6569 make this one obsolete?

**Comment by mpcomplete** [comment]
Yeah, closing this.

**Pull request closed without merging**